### PR TITLE
Implement TCP Fast Retransmit

### PR DESCRIPTION
**Pull request by barskern**
Implement TCP Fast Retransmit by simply keeping track of the **last recived ACK** (`local_rx_last_ack`) and the **number of duplicate ACKs** (`local_rx_dup_acks`) recived. When a third duplicate ACK is encountered, the `set_for_fast_retransmit` is triggered. This will make the socket send the unacknowledged packet without a delay.

A building block of this PR is an implementation of `Timer::FastRetransmit`, which provides delay-less retransmit. This was needed because `Timer::Retransmit` included a delay which is uneccessary when we recived three duplicate ACKs.

## Drawbacks

This solution is rather simple and adds some "clutter" to the `TcpSocket`. I think that preferably _TCP flow control_ would be separated into a separate state which would be a single field on the `TcpSocket`.

Closes: #104

### CODE REVIEW

File: src/socket/tcp.rs [hunk 1]
```diff
@@ -58,6 +58,9 @@ enum Timer {
     Retransmit {
         expires_at: Instant,
         delay:      Duration
+    }, 
+    FastRetransmit {
+        expires_at: Instant
```
Comment (whitequark) [inline src/socket/tcp.rs]: I think FastRetransmit doesn't need `expires_at`, since it always fires immediately.

Comment (barskern) [reply src/socket/tcp.rs]: Would returning `None` in `poll_at` imply that we want to be polled immidiately?
```rust
fn poll_at(&self) -> Option<Instant> {
    match *self {
        Timer::Idle { keep_alive_at } => keep_alive_at,
        Timer::Retransmit { expires_at, .. } => Some(expires_at),
        Timer::FastRetransmit => None,
        Timer::Close { expires_at } => Some(expires_at),
    }
}
```
https://github.com/barskern/smoltcp/blob/3d4a881b5d4129661a00d32747d6825c80531dc5/src/socket/tcp.rs#L111-L118

Comment (whitequark) [reply src/socket/tcp.rs]: No, that's an accident. The return value for "polled immediately" is `Some(Instant::from_millis(0))`. You could actually put that in `Retransmit { expires_at }`, but `FastRetransmit` is definitely cleaner.

Comment (barskern) [reply src/socket/tcp.rs]: But isn't the return value from `poll_at` supposed to be an absolute value? So we're kinda returning that it should have been polled in the future?

Comment (pothos) [reply src/socket/tcp.rs]: I'm not a reviewer, but I think I can answer your question. The return value of `poll_at` lands in `poll_delay` which is used for `wait`. `None` means to block until a packet arrives but a timeout of 0 will return immediately. `poll_delay` will do this for all return values of `poll_at` that are older than the current time given to `poll_delay`. To be sure I think you should just return 0 here as well.

Comment (dlrobertson) [reply src/socket/tcp.rs]: See the implementation for UDP socket [here]. `Some(Instant::from_millis(0))` would be some time in the past, which means poll now AFAIK.

[here]: https://github.com/m-labs/smoltcp/blob/master/src/socket/udp.rs#L215-L221

Comment (pothos) [reply src/socket/tcp.rs]: (Oh, wrote my answer before all the other comments but did not see them.)

Comment (whitequark) [reply src/socket/tcp.rs]: Do you think you can make a PR doing that?

Comment (barskern) [reply src/socket/tcp.rs]: It's on the way. I'll wait with implementing the rest of this PR until the new `poll_at` is defined.

Comment (barskern) [reply src/socket/tcp.rs]: I was thinking it would be appropriate to discuss the signature hence started #215.

**Comment by m-labs-homu** [comment]
:umbrella: The latest upstream changes (presumably 3fb0c22fd4edcecd75f4c6767a86b0333cac9366) made this pull request unmergeable. Please resolve the merge conflicts.

**Comment by barskern** [comment]
Yes will do. I normally do for my own projects I just wanted to keep history mostly intact when there are more people involved. Isn't it easier to review when there are multiple commits or should I just always squash so there is just one?

**Comment by barskern** [comment]
Updated the docstrings for `PollAt` aswell (as mentionned in the review), hence the two commits.

**Comment by whitequark** [comment]
@m-labs-homu r+

**Comment by m-labs-homu** [comment]
:pushpin: Commit 17de43d has been approved by `whitequark`

**Comment by m-labs-homu** [comment]
:hourglass: Testing commit 17de43d3eb373c3c5bcdd14e0a98aff83fc3c488 with merge 375b0266fe4a3d92ce685d4695636be63cc61d5f...

**Comment by m-labs-homu** [comment]
:sunny: Test successful - [status-travis](https://travis-ci.org/m-labs/smoltcp/builds/380301960?utm_source=github_status&utm_medium=notification)
Approved by: whitequark
Pushing 375b0266fe4a3d92ce685d4695636be63cc61d5f to master...

**Pull request closed without merging**